### PR TITLE
Speculative Decoding with NGRAM support for XPU

### PR DESCRIPTION
**Pull request by ANSHUMAN87**
This PR adds support for NGRAM on XPU target.

JH...!

---
### CI States

Latest PR Test (Base): :x: [Run #35699830673](https://github.com/sgl-project/sglang/actions/runs/35699830673)
Latest PR Test (Extra): :x: [Run #35699830563](https://github.com/sgl-project/sglang/actions/runs/35699830563)
Latest PR Test (AMD ROCm 10): :x: [Run #35699830695](https://github.com/sgl-project/sglang/actions/runs/35699830695)

### CODE REVIEW

File: python/sglang/srt/speculative/ngram_worker.py [hunk 1]
```diff
@@ -38,6 +46,111 @@
...
+    first_sibling = torch.where(
+        same_parent, j, torch.full_like(same_parent, n, dtype=torch.long)
+    ).amin(dim=1)
+    next_sibling = torch.where(
+        has_parent & (first_sibling < n),
+        first_sibling,
+        torch.full_like(first_sibling, -1),
+    )
+
+    positions.copy_(positions_2d.reshape(-1).to(positions.dtype))
```
Comment (siju-samuel) [inline python/sglang/srt/speculative/ngram_worker.py]: positions.view(bs, n).copy_(positions_2d) is this better here?

**Review by siju-samuel — CHANGES_REQUESTED** [review]
resolve merge conflicts
Zero tests for a hand-written kernel implementation
Add E2E XPU data on accuracy/accept-lentgh etc

**Comment by ANSHUMAN87** [comment]
> resolve merge conflicts Zero tests for a hand-written kernel implementation Add E2E XPU data on accuracy/accept-lentgh etc

Conflict resolved.
Now pytorch kernel is replaced with triton kernel and test case added. Please recheck.
Accuracy verification is in built to gsm8k test cases, which is passing. 

Score: 81%
Accept Length: 1.9


JH...!

File: python/sglang/srt/speculative/ngram_worker.py (hunk 1 above)
Comment (ANSHUMAN87) [reply python/sglang/srt/speculative/ngram_worker.py]: This kernel is replaced with triton.

**Comment by mingfeima** [comment]
@ANSHUMAN87 Hi is this one ready?

cc @Valentine233

**Comment by ANSHUMAN87** [comment]
> @ANSHUMAN87 Hi is this one ready?
> 
> cc @Valentine233

@mingfeima yes this PR is ready for review. Thanks!

JH...!

File: python/sglang/srt/speculative/ngram_worker.py [hunk 2]
```diff
@@ -3,11 +3,22 @@
 
 import numpy as np
 import torch
-from sgl_kernel.speculative import reconstruct_indices_from_tree_mask
+
```
Comment (Valentine233) [inline python/sglang/srt/speculative/ngram_worker.py:6]: Please add triton kernel for `reconstruct_indices_from_tree_mask` here: https://github.com/sgl-project/sglang/blob/main/python/sglang/kernels/aot/python/sgl_kernel/speculative.py#L92.

File: test/registered/xpu/test_reconstruct_tree_mask_xpu.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
+"""Correctness guard for the fused XPU tree-mask reconstruction kernel.
```
Comment (Valentine233) [inline test/registered/xpu/test_reconstruct_tree_mask_xpu.py:1]: This is to test the triton implementation of `reconstruct_indices_from_tree_mask`. It's better to move under https://github.com/sgl-project/sglang/tree/main/test/registered/spec and test all GPU devices including intel XPU.

**Review by Valentine233 — CHANGES_REQUESTED** [review]
Please provide performance result.

**Comment by Valentine233** [comment]
Hi, what is the status for this PR?

File: python/sglang/srt/speculative/ngram_worker.py (hunk 2 above)
Comment (ANSHUMAN87) [reply python/sglang/srt/speculative/ngram_worker.py:6]: [This ](https://github.com/sgl-project/sglang/blob/main/python/sglang/kernels/aot/python/sgl_kernel/speculative.py#L92.) is AOT code, so it may not be appropriate to put triton kernel hook here. Let me know your view on this.

Comment (Valentine233) [reply python/sglang/srt/speculative/ngram_worker.py:6]: According to https://github.com/sgl-project/sglang/pull/32648, I suppose it doesn't matter.

**Comment by ANSHUMAN87** [comment]
Perf Summary:


<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 15">
<meta name=Originator content="Microsoft Word 15">
<link rel=File-List
href="file:///C:/Users/anshuma2/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

<link rel=themeData
href="file:///C:/Users/anshuma2/AppData/Local/Temp/msohtmlclip1/01/clip_themedata.thmx">
<link rel=colorSchemeMapping
href="file:///C:/Users/anshuma2/AppData/Local/Temp/msohtmlclip1/01/clip_colorschememapping.xml">

<style>

</style>

</head>

<body lang=EN-US style='tab-interval:.5in;word-wrap:break-word'>


<p class=MsoNormal><b>NGRAM vs base (with corpus reset between batches):
intel_xpu vs triton, on Qwen2.5-Coder-7B-Instruct (concurrency=4, page_size=1)<o:p></o:p></b></p>


Backend | Scenario | BFS(min-max) | Match | Draft    Tokens | TPS | Accept    Len | Speedup | Status
-- | -- | -- | -- | -- | -- | -- | -- | --
intel_xpu | (base,   no-spec) | - | - | - | 111.97 | - | 1.000x | ok
intel_xpu | chain,   8 steps (BFS) | 1-1 | BFS | 8 | 180.41 | 1.855 | 1.611x | ok
intel_xpu | chain,   12 steps (BFS) | 1-1 | BFS | 12 | 163.49 | 1.855 | 1.460x | ok
intel_xpu | chain,   16 steps (BFS) | 1-1 | BFS | 16 | 162.62 | 1.855 | 1.452x | ok
intel_xpu | chain,   8 steps (PROB) | 1-1 | PROB | 8 | 180.28 | 1.855 | 1.610x | ok
intel_xpu | chain,   12 steps (PROB) | 1-1 | PROB | 12 | 163.90 | 1.855 | 1.464x | ok
intel_xpu | chain,   16 steps (PROB) | 1-1 | PROB | 16 | 163.84 | 1.855 | 1.463x | ok
triton | (base,   no-spec) | - | - | - | 110.67 | - | 1.000x | ok
triton | chain   (bfs=1) | 1-1 | BFS | 8 | 141.16 | 1.855 | 1.275x | ok
triton | narrow   BFS (bfs=2) | 1-2 | BFS | 12 | 130.82 | 1.855 | 1.182x | ok
triton | default   BFS (bfs=4) | 1-4 | BFS | 12 | 134.62 | 1.910 | 1.216x | ok
triton | wide   BFS (bfs=8) | 1-8 | BFS | 16 | 134.70 | 1.910 | 1.217x | ok
triton | default   PROB (bfs=4) | 1-4 | PROB | 12 | 130.86 | 1.855 | 1.182x | ok
triton | wide   PROB (bfs=8) | 1-8 | PROB | 16 | 130.96 | 1.855 | 1.183x | ok



<p class=MsoNormal><b>Best config per backend</b><o:p></o:p></p>

<ul style='margin-top:0in' type=disc>
 <li class=MsoNormal style='mso-list:l0 level1 lfo1;tab-stops:list .5in'>intel_xpu:&nbsp;<b>chain,
     8 steps (BFS)</b>&nbsp;(bfs=1-1, match=BFS, draft_tokens=8) —
     speedup&nbsp;<b>1.611x</b>, accept_length 1.855<o:p></o:p></li>
 <li class=MsoNormal style='mso-list:l0 level1 lfo1;tab-stops:list .5in'>triton:&nbsp;<b>chain
     (bfs=1)</b>&nbsp;(bfs=1-1, match=BFS, draft_tokens=8) — speedup&nbsp;<b>1.275x</b>,
     accept_length 1.855<o:p></o:p></li>
</ul>


</body>

</html>


JH...!

File: test/registered/xpu/test_reconstruct_tree_mask_xpu.py (hunk 3 above)
Comment (ANSHUMAN87) [reply test/registered/xpu/test_reconstruct_tree_mask_xpu.py]: moved to test/registered/kernel/speculative/test_reconstruct_tree_mask.py

Comment (ANSHUMAN87) [reply test/registered/xpu/test_reconstruct_tree_mask_xpu.py]: After moving this test file to new location, getting below lint error:
ERROR: Registered-test taxonomy violations:
  test/registered/kernel/speculative/test_reconstruct_tree_mask.py: kernel tests must use a *-kernel-* suite

I am think to move this test to manual folder. Please let me know your view.

File: python/sglang/srt/speculative/ngram_worker.py [hunk 4]
```diff
@@ -3,11 +3,22 @@
...
+try:
+    from sgl_kernel.speculative import (
+        reconstruct_indices_from_tree_mask as _reconstruct_indices_kernel,
+    )
+except ImportError:
+    # Stale / CPU-only sgl_kernel wheels may not ship this op. Fall back to the
+    # pure-torch implementation below.
+    _reconstruct_indices_kernel = None
```
Comment (Valentine233) [inline python/sglang/srt/speculative/ngram_worker.py]:
```suggestion
from sglang.srt.utils import is_xpu
if is_xpu():
    from sglang.kernels.ops.speculative.reconstruct_tree import reconstruct_indices_from_tree_mask_triton as reconstruct_indices_from_tree_mask
else:
    from sgl_kernel.speculative import reconstruct_indices_from_tree_mask
```

File: python/sglang/srt/speculative/ngram_worker.py [hunk 5]
```diff
@@ -3,11 +3,22 @@
...
 
 from sglang.kernels.ops.speculative.cache_locs import (
     assign_extend_cache_locs_func as assign_extend_cache_locs_func,
 )
+from sglang.kernels.ops.speculative.reconstruct_tree import (
+    reconstruct_indices_from_tree_mask_triton,
+)
```
Comment (Valentine233) [inline python/sglang/srt/speculative/ngram_worker.py]:
```suggestion
```

File: python/sglang/srt/speculative/ngram_worker.py [hunk 6]
```diff
@@ -72,6 +83,46 @@ def _derive_tree_links(
...
+        verified_seq_len,
+        positions,
+        retrieve_index,
+        retrieve_next_token,
+        retrieve_next_sibling,
+        batch_size,
+        draft_token_num,
+    )
+
+
```
Comment (Valentine233) [inline python/sglang/srt/speculative/ngram_worker.py]: Remove.

**Comment by Valentine233** [comment]
For perf result, why is BFS always 1-1 for intel_xpu?

**Comment by ANSHUMAN87** [comment]
> why is BFS always 1-1 for intel_xpu?

bfs>1 is equivalent to topk>1 for intel_xpu attention backend. For which currently we dont support.

**Review by Valentine233 — APPROVED** [review]

File: test/manual/kernels/test_reconstruct_tree_mask.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
+"""Test the Triton implementation of ``reconstruct_indices_from_tree_mask``.
```
Comment (mingfeima) [inline test/manual/kernels/test_reconstruct_tree_mask.py:1]: change this test file to `test/registered/kernels/ops/speculative/` and register for `register_xpu_ci`

**Review by mingfeima — APPROVED** [review]
some minor changes needed. and please rebase to master.

File: test/manual/kernels/test_reconstruct_tree_mask.py (hunk 7 above)
Comment (ANSHUMAN87) [reply test/manual/kernels/test_reconstruct_tree_mask.py:1]: Moving to registered/kernels path will cause lint error. Please refer https://github.com/sgl-project/sglang/pull/31362#discussion_r4033257175. Please let me know your view. Thanks!

**Comment by ANSHUMAN87** [comment]
> some minor changes needed. and please rebase to master.

Rebased.

File: test/manual/kernels/test_reconstruct_tree_mask.py (hunk 7 above)
Comment (ANSHUMAN87) [reply test/manual/kernels/test_reconstruct_tree_mask.py:1]: moved to test/manual/kernels/test_reconstruct_tree_mask.py now. Thanks!

**Review by mingfeima — APPROVED** [review]
